utils/pattern/src/num_pattern/mod.rs

```diff
@@ -0,0 +1,366 @@
+/// assert_writeable_eq!(
+///     pattern.interpolate(["Alice", "Bob"]),
+///     "Hello, Alice and Bob!"
+/// );
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NumericPlaceholderPattern<Store: ?Sized, const N: usize> {
+    store: Store,
+}
+
+/// An item in a [`NumericPlaceholderPattern`]. Items are either string literals or placeholders.
+#[allow(clippy::exhaustive_enums)] // documented structure
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NumericPlaceholderPatternItem<'a> {
+    /// A string literal. This can occur in one of three places:
+    ///
+    /// 1. Between the start of the string and the first placeholder (prefix)
+    /// 2. Between two placeholders (infix)
+    /// 3. Between the final placeholder and the end of the string (suffix)
+    Literal(&'a str),
+    /// A numeric placeholder. The `usize` is the placeholder number.
+    Placeholder(usize),
+}
+
+#[cfg(feature = "alloc")]
+impl<'a> From<&'a PatternToken<'_, usize>> for NumericPlaceholderPatternItem<'a> {
+    fn from(value: &'a PatternToken<usize>) -> Self {
+        match value {
+            PatternToken::Literal { content, .. } => {
+                NumericPlaceholderPatternItem::Literal(&content)
+            }
+            PatternToken::Placeholder(number) => {
+                NumericPlaceholderPatternItem::Placeholder(*number)
+            }
+        }
+    }
+}
+
+impl<Store, const N: usize> NumericPlaceholderPattern<Store, N> {
+    pub fn from_store(store: Store) -> Self {
+        Self { store }
+    }
+    pub fn take_store(self) -> Store {
+        self.store
+    }
+}
+
+impl<Store: ?Sized, const N: usize> NumericPlaceholderPattern<Store, N> {
+    pub fn borrow_store(&self) -> &Store {
+        &self.store
+    }
+}
+
+impl<Store, const N: usize> NumericPlaceholderPattern<Store, N>
+where
+    Store: ?Sized + AsRef<str>,
+{
+    /// Iterate over the contents of this pattern string, returning instances of
+    /// [`NumericPlaceholderPatternItem`].
+    pub fn iter(&self) -> impl Iterator<Item = NumericPlaceholderPatternItem> + '_ {
+        let first_placeholder_offset_char = self.store.as_ref().chars().next().unwrap_or('\0');
+        NumericPlaceholderPatternIterator::<N> {
+            pattern: self.store.as_ref(),
+            offset: first_placeholder_offset_char as usize,
+            placeholder_index: 0,
+            placeholder_position: char::len_utf8(first_placeholder_offset_char),
+        }
+    }
+
+    /// Convert this pattern's store to `&str`.
+    pub fn as_borrowed_store(&self) -> NumericPlaceholderPattern<&str, N> {
+        NumericPlaceholderPattern {
+            store: self.store.as_ref(),
+        }
+    }
+
+    pub fn byte_len(&self) -> usize {
+        self.store.as_ref().len()
+    }
+
+    /// Returns a [`Writeable`] that interpolates items from the given replacement provider
+    /// into this pattern string.
+    pub fn interpolate<'a, P>(&'a self, replacements: P) -> impl Writeable + fmt::Display + 'a
```

Comment:
This should be more strict about the number of replacements in order to remove the fallibility in line 308. I'd expect it to take `[&dyn Writeable; N]`

Reply:
Yes I will do this if we move forward with this design.

Reply:
Replaced with a trait that can be implemented on arrays of a specific length when needed.